### ISSUE TITLE
Add `getCount()`, `getCountBy($criteria)`, and `getCountBy<Field>($fieldValue)` methods to `Doctrine\ORM\EntityRepository`

### ISSUE THREAD

**Issue by kewlar** [doctrine/doctrine2#6098]
It would be convenient to have a `getCount()` and its variations `getCountBy()` and `getCountBy<Field>` methods in the `Doctrine\ORM\EntityRepository`. This functionality is often used, and is usually re-implemented. Having it in the `EntityRepository` out of the box would save time by not having to re-implement it over and over again in the inheriting repository classes.

Other explored alternatives, from the user's point of view, all have some downsides:

1. Re-implement `getCount()` and its variants in the repositories that need it. Leads to duplicated code, and wasted effort writing it.
2. Extend `EntityRepository`, and use the inheritor as the new base class for all repository classes. Prevents being able to extend another `EntityRepository` extending class (such as Gedmo extensions).
3. Implement the functionality in a `CountingRepository` trait. Having to access private (protected) `EntityRepository` members seems dirty. The trait cannot be "hinted" as meant for `EntityRepository` extending classes only, which both prevents making use of IDE and/or code analysis tools, and may lead to user mistakes.

So it seems that the best way to have `getCount()` would be to add it to the `Doctrine\ORM\EntityRepository` class. The class already has other excellent helper methods `findBy()`, `findOneBy()`, and, via the magic of `__call()`, `findBy<Field>()`, and `findOneBy<Field>()`. Adding `getCount()` seems like a natural fit.

If you think it could be useful, I could submit a PR a bit later.

**Comment by Ocramius** [comment]
Did you see #6003?

**Comment by kewlar** [comment]
This one slipped my ~~radar~~ search, sorry.

Well, it's awesome, then. Thanks!

**Comment by Ocramius** [comment]
@kewlar no problem! It was recently merged and will be in `2.6.0`, so we didn't expect you to know :-)

**Comment by dvapelnik** [comment]
I think that bad idea. Using magic methods is always wrong way because it makes unexpectable library behavior. For instance I was surprised when I have make method
```php
public function findOneByEmail($email)
{
    return $this->findOneBy(array('hash' => md5($email)));
}
```
and caught exception `Entity 'Email' has no field 'email'`. Code digging shows me method `__call()` implementation but I think that this is overehead and unexpectable behavior
Link: https://github.com/doctrine/doctrine2/blame/master/lib/Doctrine/ORM/EntityRepository.php#L224
